### PR TITLE
"Listening to" Discord RPC

### PR DESCRIPTION
**Pull request by 3top1a**
Implements "Listening to" Discord PRC in issue  #487.

Some implementation notes:
- It's absolutely hacked together, as I do not know how the code base works
- It has a git dependency for a fork of the discord-rpc with the activity field added in
- Many of the discord settings are now unnecessary (Unless you want to prefix the name's song with "Headbanging to ..."
- Because the `updateActivity` function is called every second, the time may jitter around (see gif below; discord updates the played time automatically so it only needs to be set and cleared)

anim

Screenshot_20241017_201719
image

### CODE REVIEW

**Comment by 3top1a** [comment]
Now that I'm looking at it, #454 has done almost the same thing

**Comment by Mar0xy** [comment]
> Now that I'm looking at it, #454 has done almost the same thing

I would say this supersedes #454 due to the fact that this also adds the time bar

File: src/scripts/discord.ts [hunk 1]
```diff
@@ -98,10 +103,11 @@ const getActivity = (): Presence => {
...
       const date = new Date();
-      const now = (date.getTime() / 1000) | 0;
-      const remaining = date.setSeconds(date.getSeconds() + (durationSeconds - currentSeconds));
-      presence.startTimestamp = now;
-      presence.endTimestamp = remaining;
+      const now = Math.floor(date.getTime() / 1000);
+      const startTimestamp = now - currentSeconds;
+      const endTimestamp = startTimestamp + durationSeconds;
+      presence.startTimestamp = startTimestamp;
+      presence.endTimestamp = endTimestamp;
```
Comment (Mar0xy) [inline src/scripts/discord.ts:110]: This oculd have simply been 2 lines

```js
presence.startTimestamp = now - currentSeconds;
presence.endTimestamp = startTimestamp + durationSeconds;
```

File: package.json [hunk 2]
```diff
@@ -44,7 +44,7 @@
     "@types/swagger-jsdoc": "^6.0.4",
     "axios": "^1.7.7",
     "cors": "^2.8.5",
-    "discord-rpc": "^4.0.1",
+    "discord-rpc": "github:sKiLdUsT/RPC#types",
```
Comment (Mar0xy) [inline package.json:47]: This should probably be published to NPM to avoid relying on a github repo which could vanish anytime

Comment (3top1a) [reply package.json:47]: discordjs/RPC has not been updated in 3 years, see [this issue](https://github.com/discordjs/RPC/pull/149)
Mentioned in that issue is [this PR](https://github.com/Inrixia/neptune-plugins/pull/37) that uses [@xhayper/discord-rpc](https://github.com/xhayper/discord-rpc).
I'll rewrite it to use that tomorrow.

**Comment by 3top1a** [comment]
I will do some more enhancements, however it is in a functioning state now

**Comment by 3top1a** [comment]
Because of how this works now, I've removed all redundant settings except the details prefix (includeTimestamps is now on by default because it has a sane implementation, buttons do not work and I don't believe Tidal should be announcing itself when nothing is playing, so now it's just like Spotify).
image
This makes the code much simpler, but leaves the end result untouched. I've also added a system to retry connecting to discord if it fails.
image

**Comment by Mastermindzh** [comment]
The arbitrary removal of features other users have requested & use just to make it more like Spotify is a weird move.

Please revert the unnecessary settings changes, the extra options are nice to have for users that prefer them.

Thanks.

**Comment by Mar0xy** [comment]
Buttons do actually still work just they no longer show up for your own user (you have to view it from another user)

**Comment by 3top1a** [comment]
> Buttons do actually still work just they no longer show up for your own user (you have to view it from another user)

Oh, I didn't realize that, thanks.

> The arbitrary removal of features other users have requested & use just to make it more like Spotify is a weird move.
> 
> Please revert the unnecessary settings changes, the extra options are nice to have for users that prefer them.
> 
> Thanks.

Understandable, showing 'Browsing Tidal' or turning off timestamps didn't seem useful to me so I removed them.

**Review by Mar0xy — APPROVED** [review]

**Comment by aljustiet** [comment]
https://github.com/user-attachments/assets/2606c5f1-c69c-417f-b605-fdff3f43012a

**Pull request merged**